jobs/determine-webapp-services.groovy

```diff
@@ -0,0 +1,162 @@
+//import vars.buildmaster
+//import vars.exec
+//import vars.kaGit
+//import vars.notify
+//import vars.withTimeout
+
+new Setup(steps
+
+).allowConcurrentBuilds(
+
+).addStringParam(
+    "GIT_REVISION",
+    """<b>REQUIRED</b>. The sha1 to determine the deployed services for.""",
+    ""
+
+).addStringParam(
+   "BASE_REVISION",
+   """Compute services that would be deployed by the commit since this 
+   revision based on the the commits between BASE_REVISION..GIT_REVISION.""",
+   ""
+
+).addStringParam(
+   "DEPLOYER_USERNAME",
+   """Who asked to run this job, used to ping on slack.
+If not specified, guess from the username of the person who started
+this job in Jenkins.  Typically not set manually, but by hubot scripts
+such as sun.  You can, but need not, include the leading `@`.""",
+   ""
+
+).addStringParam(
+   "SLACK_CHANNEL",
+   "The slack channel to which to send failure alerts.",
+   "#1s-and-0s-deploys"
+
+).addStringParam(
+   "SLACK_THREAD",
+   """The slack thread (must be in SLACK_CHANNEL) to which to send failure
+alerts.  By default we do not send in a thread.  Generally only set by the
+buildmaster, to the 'thread_ts' or 'timestamp' value returned by the Slack
+API.""", 
+   ""
+
+).addStringParam(
+   "REVISION_DESCRIPTION",
+   """Set by the buildmaster to give a more human-readable description
+of the GIT_REVISION, especially if it is a commit rather than a branch.""",
+   ""
+
+).addStringParam(
+   "BUILDMASTER_DEPLOY_ID",
+   """Set by the buildmaster, can be used by scripts to associate jobs
+that are part of the same deploy.  Write-only; not used by this script.""",
+   ""
+
+).addStringParam(
+  "JOB_PRIORITY",
+  """The priority of the job to be run (a lower priority means it is run
+sooner). The Priority Sorter plugin reads this parameter in to reorder jobs
+in the queue accordingly. Should be set to 3 if the job is depended on by
+the currently deploying branch, otherwise 6. Legal values are 1
+through 11. See https://jenkins.khanacademy.org/advanced-build-queue/
+for more information.""",
+  "6"
+
+).apply();
+
+currentBuild.displayName = ("${currentBuild.displayName} " +
+                            "(${params.REVISION_DESCRIPTION})");
+
+def checkArgs() {
+   if (!params.GIT_REVISION) {
+      notify.fail('The GIT_REVISION parameter is required.');
+   }
+}
+
+def checkoutWebapp() {
+   deployer_username = notify.getDeployerUsername(params.DEPLOYER_USERNAME)
+    
+   // TODO(csilvers): have these return an error message instead of alerting 
+   // themselves, so we can use notify.fail().
+   withEnv(["SLACK_CHANNEL=${params.SLACK_CHANNEL}",
+            "SLACK_THREAD=${params.SLACK_THREAD}",
+            "DEPLOYER_USERNAME=${deployer_username}",
+            "JOB_PRIORITY=${params.JOB_PRIORITY}"]) {
+      kaGit.safeSyncToOrigin("git@github.com:Khan/webapp",
+                             params.GIT_REVISION)
+   }
+}
+
+String[] determineServicesToDeploy() {
```

Comment:
Please use `def` just to be consistent with other code in this repo.  And every place you use `String` below.

Reply:
I was thinking of doing this piecemeal since typing is optional but may end up throwing errors during runtime. Changing the whole repo at once seems high risk. I'll remove the typing for now, as requested.

Reply:
I'm on board with doing it piecemeal!, but let's do it a job at a time, rather than mixing and matching within a single file.

Reply:
Ahh, well then what do you think about this being the first file? Some of the function definitions are still `def` which is  correct in the typing because they don't have a return type.

https://groovy-lang.org/objectorientation.html#_method_definition

Reply:
I've tried researching and using the groovy playground to see if it's possible to have a return type of null, for example: `Null functionName(){}`. I was not able to get syntax that didn't produce a compilation error. Groovy also implicitly returns the result of the last line executed in a function so I think the "typed function that returns nothing" should be:

```
def foo(){
    // do things
    return null
}
```